kds/kds.go

```diff
@@ -115,19 +133,94 @@ var (
 )
 
 // TCBVersion is a 64-bit bitfield of different security patch levels of AMD firmware and microcode.
+//
+// Deprecated: use TCBVersionStruct instead
 type TCBVersion uint64
 
+// TCBVersionStruct represents the SEV-SNP TCB_VERSION
+type TCBVersionStruct struct {
+	version uint8
+	// TCB is a 64-bit bitfield of different security patch levels of AMD firmware and microcode.
+	TCB uint64
+}
+
+// String returns the version of the TCB in string format
+func (v TCBVersionStruct) String() string {
+	switch v.version {
+	case tcbStructVersion0:
+		return fmt.Sprintf("Milan or Genoa TCB: 0x%x", v.TCB)
+	case tcbStructVersion1:
+		return fmt.Sprintf("Turin TCB: 0x%x", v.TCB)
+	default:
+		return "Unknown TCB"
+	}
+}
+
+// FromTCBVersion initializes TCBVersionStruct from TCBVersion
+func (v *TCBVersionStruct) FromTCBVersion(tcb TCBVersion) {
+	v.version = tcbStructVersion0
+	v.TCB = uint64(tcb)
+}
+
+// ToTCBVersion converts the object to TCBVersion
+func (v TCBVersionStruct) ToTCBVersion() TCBVersion {
+	return TCBVersion(v.TCB)
+}
+
+// ToTCBParts converts the object to TCBParts
+func (v TCBVersionStruct) ToTCBParts() (TCBParts, error) {
+	switch v.version {
+	case tcbStructVersion1:
+		return TCBParts{
+			version:  v.version,
+			UcodeSpl: uint8((v.TCB >> 56) & 0xff),
+			SnpSpl:   uint8((v.TCB >> 24) & 0xff),
+			TeeSpl:   uint8((v.TCB >> 16) & 0xff),
+			BlSpl:    uint8((v.TCB >> 8) & 0xff),
+			FmcSpl:   uint8((v.TCB >> 0) & 0xff),
+		}, nil
+	case tcbStructVersion0:
+		return TCBParts{
+			version:  v.version,
+			UcodeSpl: uint8((v.TCB >> 56) & 0xff),
+			SnpSpl:   uint8((v.TCB >> 48) & 0xff),
+			Spl7:     uint8((v.TCB >> 40) & 0xff),
+			Spl6:     uint8((v.TCB >> 32) & 0xff),
+			Spl5:     uint8((v.TCB >> 24) & 0xff),
+			Spl4:     uint8((v.TCB >> 16) & 0xff),
+			TeeSpl:   uint8((v.TCB >> 8) & 0xff),
+			BlSpl:    uint8((v.TCB >> 0) & 0xff),
+		}, nil
+	default:
+		return TCBParts{}, fmt.Errorf("unknown TCB version: %d", v.version)
+	}
+}
+
+// NewTCBVersionStruct creates new TCBVersionStruct from product line and 64-bit TCB value
+func NewTCBVersionStruct(productLine string, tcb uint64) (*TCBVersionStruct, error) {
+	c := cases.Title(language.English)
+	switch c.String(productLine) {
+	case "Turin":
+		return &TCBVersionStruct{version: tcbStructVersion1, TCB: tcb}, nil
+	default:
```

Comment:
This should fail closed. We don't want "Venice" to accidentally end up misinterpreted here.

Reply:
@deeglaze When I do that, some of the validate tests are failing. The root cause is that the Reports generated by the QuoteProvider doesn't have the FMS set. Do you know if we could configure the QuoteProvider to fill the FMS in the report?

Reply:
The issue is the report generated in the following code doesn't have FMS in the report:
```go
		{
			name: "just reportData",
			attestation: func() *spb.Attestation {
				q, err := sg.GetQuoteProto(qp0, nonce0s1)
				if err != nil {
					t.Fatal(err)
				}
				report := q.Report
				//report.Cpuid1EaxFms = 0xa00f11
				return &spb.Attestation{
					Report: report,
					CertificateChain: &spb.CertificateChain{
						AskCert:  sign0.Ask.Raw,
						ArkCert:  sign0.Ark.Raw,
						VcekCert: sign0.Vcek.Raw,
					},
				}
			}(),
			opts: &Options{ReportData: nonce0s1[:], GuestPolicy: abi.SnpPolicy{Debug: true}},
		},
```

Reply:
Yes, we can use test data that depends on version 3 and populates the FMS.

Reply:
I could add file and track an issue to use v3 reports in tools/check and validate_test, if you're OK with defaulting to structVersion0 here.